controller/resource.go

```diff
@@ -0,0 +1,240 @@
+package controller
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/aws/aws-sdk-go/service/ecs/ecsiface"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/aws/aws-sdk-go/service/secretsmanager/secretsmanageriface"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/go-hclog"
+)
+
+const meshTag = "consul.hashicorp.com/mesh"
+
+// ResourceLister is an interface for listing Resources.
+type ResourceLister interface {
+	List() ([]Resource, error)
+}
+
+// Resource is a generic type that needs to be reconciled by the Controller.
+// It offers Upsert and Delete functions to reconcile itself with an external state.
+type Resource interface {
+	ID() (string, error)
+	Upsert() error
+	Delete() error
+}
+
+// TaskLister is an implementation of ResourceLister that lists ECS tasks.
+type TaskLister struct {
+	// ECSClient is the AWS ECS client to be used by the TaskLister.
+	ECSClient ecsiface.ECSAPI
+	// SecretsManagerClient is the AWS Secrets Manager client to be used by the TaskLister.
+	SecretsManagerClient secretsmanageriface.SecretsManagerAPI
+	// ConsulClient is the Consul client to be used by the TaskLister.
+	// TaskLister doesn't need to talk to Consul, but it passes this client
+	// to each Resource it creates.
+	ConsulClient *api.Client
+
+	// Cluster is the name or the ARN of the ECS cluster.
+	Cluster string
+	// SecretPrefix is the prefix to determine names of resources in Consul or AWS.
+	SecretPrefix string
+
+	// Log is the logger for the TaskLister.
+	Log hclog.Logger
+}
+
+// List lists all tasks for the Cluster.
+func (t TaskLister) List() ([]Resource, error) {
+	var resources []Resource
+	// nextToken is to handle paginated responses from AWS.
+	var nextToken *string
+
+	// This isn't an infinite loop, instead this is a "do while" loop
+	// because we'll break out of it as soon as nextToken is nil.
+	for {
+		taskListOutput, err := t.ECSClient.ListTasks(&ecs.ListTasksInput{
+			Cluster:   aws.String(t.Cluster),
+			NextToken: nextToken,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("listing tasks: %w", err)
+		}
+		nextToken = taskListOutput.NextToken
+
+		tasks, err := t.ECSClient.DescribeTasks(&ecs.DescribeTasksInput{
+			Cluster: aws.String(t.Cluster),
+			Tasks:   taskListOutput.TaskArns,
+			Include: []*string{aws.String("TAGS")},
+		})
+		if err != nil {
+			return nil, fmt.Errorf("describing tasks: %w", err)
+		}
+		for _, task := range tasks.Tasks {
+			resources = append(resources, &Task{
```

Comment:
should we check here if `task` is nil and if it is error and if it's not dereference it so we don't have to worry about nil pointers in the `Task` struct?